scala.MatchError: Quantified resource might not be injective

The following input:
```java
class C {
  /*@
  requires arr != null;
  requires Value(arr[*]);
  requires (\forall* int k = 0 .. arr.length; Value(arr[k].arr[*]));
  pure int mySum2(D[] arr) = 0;

  requires arr != null;
  requires Value(arr[*]);
  requires (\forall int k = 0 .. arr.length; arr[k].arr != null);
  pure int mySum(D[] arr) = mySum2(arr);
  @*/
}

class D {
  final int[] arr;
}
```

Gives the following output:

```
[ERROR] Unrecoverable error: Quantified resource aloc(opt_get1(arr(aloc(opt_get1(arr1), k).ref)), i).int might not be injective. (of class viper.silver.verifier.reasons$QPAssertionNotInjective)
scala.MatchError: Quantified resource aloc(opt_get1(arr(aloc(opt_get1(arr1), k).ref)), i).int might not be injective. (of class viper.silver.verifier.reasons$QPAssertionNotInjective)
        at viper.api.backend.SilverBackend.getFailure(SilverBackend.scala:279)
        at viper.api.backend.SilverBackend.getFailure$(SilverBackend.scala:279)
        at viper.api.backend.silicon.Silicon.getFailure(Silicon.scala:41)
        at viper.api.backend.SilverBackend.processError(SilverBackend.scala:143)
        at viper.api.backend.SilverBackend.processError$(SilverBackend.scala:118)
        at viper.api.backend.silicon.Silicon.processError(Silicon.scala:41)
        at viper.api.backend.SilverBackend.$anonfun$submit$9(SilverBackend.scala:108)
        at viper.api.backend.SilverBackend.$anonfun$submit$9$adapted(SilverBackend.scala:108)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at viper.api.backend.SilverBackend.submit(SilverBackend.scala:108)
        at viper.api.backend.SilverBackend.submit$(SilverBackend.scala:59)
        at viper.api.backend.silicon.Silicon.submit(Silicon.scala:41)
        at vct.main.stages.SilverBackend.verify(Backend.scala:115)
        ...
Exception in thread "main" scala.MatchError: Quantified resource aloc(opt_get1(arr(aloc(opt_get1(arr1), k).ref)), i).int might not be injective. (of class viper.silver.verifier.reasons$QPAssertionNotInjective)
        at viper.api.backend.SilverBackend.getFailure(SilverBackend.scala:279)
        at viper.api.backend.SilverBackend.getFailure$(SilverBackend.scala:279)
        at viper.api.backend.silicon.Silicon.getFailure(Silicon.scala:41)
        at viper.api.backend.SilverBackend.processError(SilverBackend.scala:143)
        at viper.api.backend.SilverBackend.processError$(SilverBackend.scala:118)
        at viper.api.backend.silicon.Silicon.processError(Silicon.scala:41)
        at viper.api.backend.SilverBackend.$anonfun$submit$9(SilverBackend.scala:108)
        at viper.api.backend.SilverBackend.$anonfun$submit$9$adapted(SilverBackend.scala:108)
        at scala.collection.immutable.List.foreach(List.scala:333)
        at viper.api.backend.SilverBackend.submit(SilverBackend.scala:108)
        at viper.api.backend.SilverBackend.submit$(SilverBackend.scala:59)
        at viper.api.backend.silicon.Silicon.submit(Silicon.scala:41)
        at vct.main.stages.SilverBackend.verify(Backend.scala:115)
        ...
```

--- a/src/col/vct/col/ast/Node.scala
+++ b/src/col/vct/col/ast/Node.scala
@@ -1590,11 +1590,11 @@ final case class DerefHeapVariable[G](ref: Ref[G, HeapVariable[G]])(
 )(implicit val o: Origin)
     extends Expr[G] with HeapDeref[G] with DerefHeapVariableImpl[G]
 final case class Deref[G](obj: Expr[G], ref: Ref[G, InstanceField[G]])(
-    val blame: Blame[InsufficientPermission]
+    val blame: Blame[ClassDerefError]
 )(implicit val o: Origin)
     extends Expr[G] with HeapDeref[G] with DerefImpl[G]
 final case class ModelDeref[G](obj: Expr[G], ref: Ref[G, ModelField[G]])(
-    val blame: Blame[ModelInsufficientPermission]
+    val blame: Blame[ClassDerefError]
 )(implicit val o: Origin)
     extends Expr[G] with ModelDerefImpl[G]
 final case class DerefPointer[G](pointer: Expr[G])(
@@ -3394,7 +3394,7 @@ sealed trait CPPExpr[G] extends Expr[G] with CPPExprImpl[G]
 final case class CPPLocal[G](
     name: String,
     genericArgs: Seq[CPPExprOrTypeSpecifier[G]],
-)(val blame: Blame[DerefInsufficientPermission])(implicit val o: Origin)
+)(val blame: Blame[ClassDerefError])(implicit val o: Origin)
     extends CPPExpr[G] with CPPLocalImpl[G] {
   var ref: Option[CPPNameTarget[G]] = None
 }
@@ -3694,10 +3694,9 @@ final case class JavaWildcard[G]()(implicit val o: Origin = DiagnosticOrigin)
     extends JavaType[G] with JavaWildcardImpl[G]
 
 sealed trait JavaExpr[G] extends Expr[G] with JavaExprImpl[G]
-final case class JavaLocal[G](name: String)(
-    val blame: Blame[DerefInsufficientPermission]
-)(implicit val o: Origin)
-    extends JavaExpr[G] with JavaLocalImpl[G] {
+final case class JavaLocal[G](name: String)(val blame: Blame[ClassDerefError])(
+    implicit val o: Origin
+) extends JavaExpr[G] with JavaLocalImpl[G] {
   var ref: Option[JavaNameTarget[G]] = None
 }
 final case class JavaDeref[G](obj: Expr[G], field: String)(
@@ -4102,7 +4101,7 @@ final case class LLVMExtractValue[G](
     resultType: Type[G],
     value: Expr[G],
     indices: Seq[Int],
-)(val blame: Blame[InsufficientPermission])(implicit val o: Origin)
+)(val blame: Blame[ClassDerefError])(implicit val o: Origin)
     extends LLVMExpr[G] with LLVMExtractValueImpl[G]
 
 final case class LLVMSignExtend[G](
@@ -4365,10 +4364,9 @@ final case class PVLNamedType[G](name: String, typeArgs: Seq[Type[G]])(
 }
 
 sealed trait PVLExpr[G] extends Expr[G] with PVLExprImpl[G]
-final case class PVLLocal[G](name: String)(
-    val blame: Blame[DerefInsufficientPermission]
-)(implicit val o: Origin)
-    extends PVLExpr[G] with PVLLocalImpl[G] {
+final case class PVLLocal[G](name: String)(val blame: Blame[ClassDerefError])(
+    implicit val o: Origin
+) extends PVLExpr[G] with PVLLocalImpl[G] {
   var ref: Option[PVLNameTarget[G]] = None
 }
 final case class PVLDeref[G](obj: Expr[G], field: String)(

--- a/src/col/vct/col/origin/Blame.scala
+++ b/src/col/vct/col/origin/Blame.scala
@@ -33,6 +33,14 @@ case class NegativePermissionValue(node: Expr[_]) extends ContractFailure {
   override def inlineDescCompletion: String =
     s"${node.o.inlineContextText} may be a negative permission amount"
 }
+case class NotWellDefined(node: Node[_], inner: VerificationFailure)
+    extends ContractFailure {
+  override def code: String = inner.code
+  override def descCompletion =
+    s"the contract is not well-defined:\n${inner.desc}"
+  override def inlineDescCompletion =
+    s"in ${node.o.inlineContextText} the contract is not well-defined: ${inner.inlineDesc}"
+}
 
 trait VerificationFailure {
   def code: String

--- a/src/col/vct/col/origin/Blame.scala
+++ b/src/col/vct/col/origin/Blame.scala
@@ -704,7 +704,15 @@ case class ChannelInvariantNotEstablished(
     s"The channel invariant at `$node` cannot be established, since $failure"
 }
 
-sealed trait DerefInsufficientPermission extends FrontendDerefError
+sealed trait ClassDerefError extends FrontendDerefError
+case class ClassNull(node: HeapDeref[_])
+    extends ClassDerefError with NodeVerificationFailure {
+  override def code: String = "classNull"
+  override def descInContext: String = "This class may be null."
+  override def inlineDescWithSource(source: String): String =
+    s"This class may be null: `$source`."
+}
+sealed trait DerefInsufficientPermission extends ClassDerefError
 case class InsufficientPermission(node: HeapDeref[_])
     extends DerefInsufficientPermission with NodeVerificationFailure {
   override def code: String = "perm"

--- a/src/rewrite/vct/rewrite/CheckProcessAlgebra.scala
+++ b/src/rewrite/vct/rewrite/CheckProcessAlgebra.scala
@@ -39,8 +39,8 @@ case class CheckProcessAlgebra[Pre <: Generation]()
   }
 
   case class InsufficientPermissionForModelField(modelDeref: ModelDeref[_])
-      extends Blame[InsufficientPermission] {
-    override def blame(error: InsufficientPermission): Unit =
+      extends Blame[ClassDerefError] {
+    override def blame(error: ClassDerefError): Unit =
       modelDeref.blame.blame(ModelInsufficientPermission(modelDeref))
   }
 

--- a/src/rewrite/vct/rewrite/ConstantifyFinalFields.scala
+++ b/src/rewrite/vct/rewrite/ConstantifyFinalFields.scala
@@ -4,10 +4,17 @@ import hre.util.ScopedStack
 import vct.col.ast._
 import vct.col.rewrite.{Generation, Rewriter, RewriterBuilder}
 import vct.col.util.AstBuildHelpers._
-import vct.col.ast.RewriteHelpers._
-import vct.col.origin.{AbstractApplicable, Origin, PanicBlame, TrueSatisfiable}
+import vct.col.origin.{
+  AbstractApplicable,
+  Blame,
+  ClassDerefError,
+  ClassNull,
+  InvocationFailure,
+  Origin,
+  PanicBlame,
+  TrueSatisfiable,
+}
 import vct.col.ref.Ref
-import vct.col.rewrite.ConstantifyFinalFields.FinalFieldPerm
 import vct.col.util.SuccessionMap
 import vct.result.VerificationError.UserError
 
@@ -16,16 +23,26 @@ case object ConstantifyFinalFields extends RewriterBuilder {
   override def desc: String =
     "Encode final fields with functions, so that they are not on the heap."
 
-  case class FinalFieldPerm(loc: FieldLocation[_]) extends UserError {
+  private case class FinalFieldPerm(loc: FieldLocation[_]) extends UserError {
     override def code: String = "finalFieldPerm"
     override def text: String =
       loc.o.messageInContext(
         "Specifying permission over final fields is not allowed, since they are treated as constants."
       )
   }
+
+  private case class FieldClassNull(
+      blame: Blame[ClassDerefError],
+      deref: HeapDeref[_],
+  ) extends Blame[InvocationFailure] {
+    override def blame(error: InvocationFailure): Unit =
+      blame.blame(ClassNull(deref))
+  }
 }
 
 case class ConstantifyFinalFields[Pre <: Generation]() extends Rewriter[Pre] {
+  import vct.col.rewrite.ConstantifyFinalFields._
+
   val currentClass: ScopedStack[Class[Pre]] = ScopedStack()
   var finalValueMap: Map[Declaration[Pre], Expr[Pre]] = Map()
   val fieldFunction: SuccessionMap[InstanceField[Pre], Function[Post]] =
@@ -75,76 +92,68 @@ case class ConstantifyFinalFields[Pre <: Generation]() extends Rewriter[Pre] {
 
   override def dispatch(decl: Declaration[Pre]): Unit =
     decl match {
-      case cls: Class[Pre] => currentClass.having(cls) { rewriteDefault(cls) }
-      case field: InstanceField[Pre] =>
+      case cls: Class[Pre] => currentClass.having(cls) { super.dispatch(cls) }
+      case field: InstanceField[Pre] if isFinal(field) =>
         implicit val o: Origin = field.o
-        if (isFinal(field)) {
-          val `this` =
-            new Variable(dispatch(
-              currentClass.top.classType(currentClass.top.typeArgs.map {
-                v: Variable[Pre] => TVar(v.ref)
-              })
-            ))
-          fieldFunction(field) = globalDeclarations
-            .declare(withResult((result: Result[Post]) =>
-              function[Post](
-                blame = AbstractApplicable,
-                contractBlame = TrueSatisfiable,
-                returnType = dispatch(field.t),
-                args = Seq(`this`),
-                requires = UnitAccountedPredicate(`this`.get !== Null()),
-                ensures = UnitAccountedPredicate(
-                  finalValueMap.get(field) match {
-                    case Some(value) =>
-                      result === substituteThisObject.having(`this`.get) {
-                        rewriteDefault(value)
-                      }
-                    case None => tt[Post]
+        val `this` =
+          new Variable(dispatch(
+            currentClass.top.classType(currentClass.top.typeArgs.map {
+              v: Variable[Pre] => TVar(v.ref)
+            })
+          ))
+        fieldFunction(field) = globalDeclarations
+          .declare(withResult((result: Result[Post]) =>
+            function[Post](
+              blame = AbstractApplicable,
+              contractBlame = TrueSatisfiable,
+              returnType = dispatch(field.t),
+              args = Seq(`this`),
+              requires = UnitAccountedPredicate(`this`.get !== Null()),
+              ensures = UnitAccountedPredicate(finalValueMap.get(field) match {
+                case Some(value) =>
+                  result === substituteThisObject.having(`this`.get) {
+                    super.dispatch(value)
                   }
-                ),
-              )
-            ))
-        } else { rewriteDefault(field) }
-      case other => rewriteDefault(other)
+                case None => tt[Post]
+              }),
+            )
+          ))
+      case other => super.dispatch(other)
     }
 
   override def dispatch(e: Expr[Pre]): Expr[Post] =
     e match {
       case ThisObject(_) if substituteThisObject.nonEmpty =>
         substituteThisObject.top
-      case Deref(obj, Ref(field)) =>
+      case d @ Deref(obj, Ref(field)) if isFinal(field) =>
         implicit val o: Origin = e.o
-        if (isFinal(field))
-          FunctionInvocation[Post](
-            fieldFunction.ref(field),
-            Seq(dispatch(obj)),
-            Nil,
-            Nil,
-            Nil,
-          )(PanicBlame("requires nothing"))
-        else
-          rewriteDefault(e)
-      case other => rewriteDefault(other)
+        functionInvocation[Post](
+          FieldClassNull(d.blame, d),
+          fieldFunction.ref(field),
+          Seq(dispatch(obj)),
+        )
+      case _ => super.dispatch(e)
     }
 
   override def dispatch(location: Location[Pre]): Location[Post] =
     location match {
       case loc @ FieldLocation(_, Ref(field)) if isFinal(field) =>
         throw FinalFieldPerm(loc)
-      case other => rewriteDefault(other)
+      case _ => super.dispatch(location)
     }
 
-  def makeInhale(obj: Expr[Pre], field: InstanceField[Pre], value: Expr[Pre])(
-      implicit o: Origin
-  ): Statement[Post] =
+  def makeInhale(
+      obj: Expr[Pre],
+      field: InstanceField[Pre],
+      value: Expr[Pre],
+      deref: Deref[Pre],
+  )(implicit o: Origin): Statement[Post] =
     Assume(
-      FunctionInvocation[Post](
+      functionInvocation[Post](
+        FieldClassNull(deref.blame, deref),
         fieldFunction.ref(field),
         Seq(dispatch(obj)),
-        Nil,
-        Nil,
-        Nil,
-      )(PanicBlame("requires nothing")) === dispatch(value)
+      ) === dispatch(value)
     )
 
   override def dispatch(stat: Statement[Pre]): Statement[Post] =
@@ -155,11 +164,11 @@ case class ConstantifyFinalFields[Pre <: Generation]() extends Rewriter[Pre] {
       case Eval(PreAssignExpression(Deref(obj, Ref(field)), value))
           if isFinal(field) && finalValueMap.contains(field) =>
         Block(Nil)(stat.o)
-      case Assign(Deref(obj, Ref(field)), value) if isFinal(field) =>
-        makeInhale(obj, field, value)(stat.o)
-      case Eval(PreAssignExpression(Deref(obj, Ref(field)), value))
+      case Assign(d @ Deref(obj, Ref(field)), value) if isFinal(field) =>
+        makeInhale(obj, field, value, d)(stat.o)
+      case Eval(PreAssignExpression(d @ Deref(obj, Ref(field)), value))
           if isFinal(field) =>
-        makeInhale(obj, field, value)(stat.o)
-      case other => rewriteDefault(other)
+        makeInhale(obj, field, value, d)(stat.o)
+      case _ => super.dispatch(stat)
     }
 }

--- a/src/rewrite/vct/rewrite/EncodeByValueClassUsage.scala
+++ b/src/rewrite/vct/rewrite/EncodeByValueClassUsage.scala
@@ -21,8 +21,8 @@ case object EncodeByValueClassUsage extends RewriterBuilder {
       assign: Node[_],
       clazz: ByValueClass[_],
       field: InstanceField[_],
-  ) extends Blame[InsufficientPermission] {
-    override def blame(error: InsufficientPermission): Unit = {
+  ) extends Blame[ClassDerefError] {
+    override def blame(error: ClassDerefError): Unit = {
       if (blame.isInstanceOf[PanicBlame]) {
         assign.o
           .blame(CopyClassFailed(assign, clazz, Referrable.originName(field)))
@@ -38,8 +38,8 @@ case object EncodeByValueClassUsage extends RewriterBuilder {
       inv: Invocation[_],
       clazz: ByValueClass[_],
       field: InstanceField[_],
-  ) extends Blame[InsufficientPermission] {
-    override def blame(error: InsufficientPermission): Unit = {
+  ) extends Blame[ClassDerefError] {
+    override def blame(error: ClassDerefError): Unit = {
       blame.blame(
         CopyClassFailedBeforeCall(inv, clazz, Referrable.originName(field))
       )
@@ -121,7 +121,7 @@ case class EncodeByValueClassUsage[Pre <: Generation]() extends Rewriter[Pre] {
       obj: Expr[Post],
       t: TByValueClass[Pre],
       target: Expr[Post],
-      blame: InstanceField[Pre] => Blame[InsufficientPermission],
+      blame: InstanceField[Pre] => Blame[ClassDerefError],
   ): Statement[Post] = {
     implicit val o: Origin = obj.o
     val ov = new Variable[Post](dispatch(t))(o.where(name = "original"))

--- a/src/rewrite/vct/rewrite/lang/LangCPPToCol.scala
+++ b/src/rewrite/vct/rewrite/lang/LangCPPToCol.scala
@@ -419,7 +419,7 @@ case object LangCPPToCol {
 
   private case class SYCLAccessorFieldInsufficientReferencePermissionBlame(
       local: CPPLocal[_]
-  ) extends Blame[InsufficientPermission] {
+  ) extends Blame[ClassDerefError] {
     private case class SYCLAccessorFieldInsufficientReferencePermissionError(
         error: InsufficientPermission
     ) extends UserError {
@@ -428,14 +428,21 @@ case object LangCPPToCol {
       override def text: String = local.o.messageInContext(error.descInContext)
     }
 
-    override def blame(error: InsufficientPermission): Unit =
-      throw SYCLAccessorFieldInsufficientReferencePermissionError(error)
+    override def blame(error: ClassDerefError): Unit = {
+      error match {
+        case permission: InsufficientPermission =>
+          throw SYCLAccessorFieldInsufficientReferencePermissionError(
+            permission
+          )
+        case ClassNull(_) => ???
+      }
+    }
   }
 
   private case class SYCLAccessorRangeIndexFieldInsufficientReferencePermissionBlame(
       inv: CPPInvocation[_]
-  ) extends Blame[InsufficientPermission] {
-    override def blame(error: InsufficientPermission): Unit =
+  ) extends Blame[ClassDerefError] {
+    override def blame(error: ClassDerefError): Unit =
       PanicBlame(inv.o.messageInContext(
         s"There was not enough permission to access" +
           s" a field containing the size of an accessor dimension. This should not be possible."

--- a/src/rewrite/vct/rewrite/lang/LangLLVMToCol.scala
+++ b/src/rewrite/vct/rewrite/lang/LangLLVMToCol.scala
@@ -121,9 +121,13 @@ case object LangLLVMToCol {
 
   private final case class PointerSubscriptToInsufficientPermissionBlame(
       blame: Blame[PointerSubscriptError]
-  ) extends Blame[InsufficientPermission] {
-    override def blame(error: InsufficientPermission): Unit =
-      blame.blame(PointerInsufficientPermission(error.node))
+  ) extends Blame[ClassDerefError] {
+    override def blame(error: ClassDerefError): Unit = {
+      blame.blame(PointerInsufficientPermission(error match {
+        case ClassNull(node) => node
+        case InsufficientPermission(node) => node
+      }))
+    }
   }
 
   private val pallasResArgPermOrigin: Origin = Origin(Seq(
@@ -1136,7 +1140,7 @@ case class LangLLVMToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
       pointer: Expr[Post],
       t: Type[Pre],
       indices: Seq[Expr[Pre]],
-      blame: Blame[InsufficientPermission],
+      blame: Blame[ClassDerefError],
   )(implicit o: Origin): Expr[Post] = {
     if (indices.isEmpty) { return pointer }
     t match {
@@ -1350,7 +1354,7 @@ case class LangLLVMToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
       value: Expr[Post],
       t: Type[Pre],
       indices: Seq[Int],
-      blame: Blame[InsufficientPermission],
+      blame: Blame[ClassDerefError],
   )(implicit o: Origin): Expr[Post] = {
     if (indices.isEmpty) { return value }
     t match {

--- a/src/rewrite/vct/rewrite/lang/LangSpecificToCol.scala
+++ b/src/rewrite/vct/rewrite/lang/LangSpecificToCol.scala
@@ -112,7 +112,7 @@ case class LangSpecificToCol[Pre <: Generation](
   def specLocal(
       target: SpecNameTarget[Pre],
       e: Expr[Pre],
-      blame: Blame[DerefInsufficientPermission],
+      blame: Blame[ClassDerefError],
   ): Expr[Post] =
     target match {
       case RefAxiomaticDataType(_) => throw NotAValue(e)
@@ -129,7 +129,7 @@ case class LangSpecificToCol[Pre <: Generation](
       obj: Expr[Pre],
       target: SpecDerefTarget[Pre],
       e: Expr[Pre],
-      blame: Blame[DerefInsufficientPermission],
+      blame: Blame[ClassDerefError],
   ): Expr[Post] =
     target match {
       case RefEnumConstant(enum, decl) =>

--- a/src/viper/viper/api/backend/SilverBackend.scala
+++ b/src/viper/viper/api/backend/SilverBackend.scala
@@ -413,6 +413,40 @@ trait SilverBackend
           .NegativePermissionValue(
             info(p).permissionValuePermissionNode.get
           ) // need to fetch access
+      // Keep in sync with defer()
+      case reasons.DivisionByZero(e) =>
+        val division = info(e).dividingExpr.get
+        blame.NotWellDefined(division, blame.ScalarDivByZero(division))
+      case reasons.InsufficientPermission(f @ silver.FieldAccess(_, _)) =>
+        val deref = get[col.SilverDeref[_]](f)
+        blame.NotWellDefined(deref, blame.InsufficientPermission(deref))
+      case reasons.InsufficientPermission(p @ silver.PredicateAccess(_, _)) =>
+        val unfolding = info(p).unfolding.get
+        blame.NotWellDefined(unfolding, blame.UnfoldFailed(unfolding))
+      case reasons.QPAssertionNotInjective(access: silver.ResourceAccess) =>
+        val starall = info(access).starall.get
+        blame.NotWellDefined(
+          starall,
+          blame.ReceiverNotInjective(starall, get(access)),
+        )
+      case reasons.LabelledStateNotReached(expr) =>
+        val old = get[col.Old[_]](expr)
+        blame.NotWellDefined(old, blame.LabelNotReached(old))
+      case reasons.SeqIndexNegative(_, idx) =>
+        val subscript = info(idx).seqIndexSubscriptNode.get
+        blame.NotWellDefined(subscript, blame.SeqBoundNegative(subscript))
+      case reasons.SeqIndexExceedsLength(_, idx) =>
+        val subscript = info(idx).seqIndexSubscriptNode.get
+        blame.NotWellDefined(subscript, blame.SeqBoundExceedsLength(subscript))
+      case reasons.MapKeyNotContained(_, key) =>
+        val get = info(key).mapGet.get
+        blame.NotWellDefined(get, blame.MapKeyError(get))
+      case reasons.AssertionFalse(expr) =>
+        val asserting = info(expr).asserting.get
+        blame.NotWellDefined(
+          asserting,
+          blame.AssertFailed(getFailure(reason), asserting),
+        )
       case r => throw new NotImplementedError("Missing: " + r)
     }
 
@@ -462,6 +496,7 @@ trait SilverBackend
         )
     }
 
+  // Keep in sync with getFailure
   def defer(reason: ErrorReason): Unit =
     reason match {
       case reasons.DivisionByZero(e) =>

